sklearn/feature_selection/tests/test_from_model.py

```diff
@@ -63,6 +65,117 @@ def test_input_estimator_unchanged():
+    def __init__(self, importances):
+        self.importances = importances
+
+    def fit(self, X, y=None):
+        self.feature_importances_ = np.array(self.importances)
+
+
+def check_max_features(est, X, y):
+    X = X.copy()
+    max_features = X.shape[1]
+
+    check_valid_max_features(est, X, y)
+    check_invalid_max_features(est, X, y)
+
+    transformer1 = SelectFromModel(estimator=est, max_features='all')
+    transformer2 = SelectFromModel(estimator=est,
+                                   max_features=max_features)
+    X_new1 = transformer1.fit_transform(X, y)
+    X_new2 = transformer2.fit_transform(X, y)
+    assert_array_equal(X_new1, X_new2)
+
+    # Test max_features against actual model.
+    transformer1 = SelectFromModel(estimator=Lasso(alpha=0.025))
+    X_new1 = transformer1.fit_transform(X, y)
+    scores1 = np.abs(transformer1.estimator_.coef_)
+    candidate_indices1 = np.argsort(-scores1, kind='mergesort')
+
+    for n_features in range(1, X_new1.shape[1] + 1):
+        transformer2 = SelectFromModel(estimator=Lasso(alpha=0.025),
+                                       max_features=n_features)
+        X_new2 = transformer2.fit_transform(X, y)
+        scores2 = np.abs(transformer2.estimator_.coef_)
+        candidate_indices2 = np.argsort(-scores2, kind='mergesort')
+        assert_array_equal(X[:, candidate_indices1[:n_features]],
+                           X[:, candidate_indices2[:n_features]])
+    assert_array_equal(transformer1.estimator_.coef_,
+                       transformer2.estimator_.coef_)
+
+    # Test if max_features can break tie among feature importance
+
+    feature_importances = np.array([4, 4, 4, 4, 3, 3, 3, 2, 2, 1])
+    for n_features in range(1, max_features + 1):
+        transformer = SelectFromModel(
+            FixedImportanceEstimator(feature_importances),
+            max_features=n_features)
+        X_new = transformer.fit_transform(X, y)
+        selected_feature_indices = np.where(transformer._get_support_mask())[0]
+        assert_array_equal(selected_feature_indices, np.arange(n_features))
+        assert_equal(X_new.shape[1], n_features)
+
+
+def check_threshold_and_max_features(est, X, y):
+    transformer1 = SelectFromModel(estimator=est, max_features=3)
+    X_new1 = transformer1.fit_transform(X, y)
+
+    transformer2 = SelectFromModel(estimator=est, threshold=0.04)
+    X_new2 = transformer2.fit_transform(X, y)
+
+    transformer3 = SelectFromModel(estimator=est, max_features=3,
+                                   threshold=0.04)
+    X_new3 = transformer3.fit_transform(X, y)
+    assert_equal(X_new3.shape[1], min(X_new1.shape[1], X_new2.shape[1]))
+    selected_indices = \
+        transformer3.transform(np.arange(X.shape[1]))[np.newaxis, :]
+    assert_array_equal(X_new3, X[:, selected_indices[0][0]])
+
+    """
+    If threshold and max_features are not provided, all features are
+    returned, use threshold=None if it is not required.
```

Comment:
and update here. `threshold=None` has a special meaning. It does not mean "infinitely low threshold"